File: app/src/main/java/presenter/task/AILoaderTask.java
package presenter.task;

import android.support.design.widget.Snackbar;
import android.view.View;
import android.widget.Button;
import model.ai.PlayerLogic;
import android.content.Context;
import android.os.AsyncTask;
import hu.bme.aut.android.chainreaction.R;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.util.ModelSerializer;

import java.io.IOException;
import java.io.InputStream;

/**
 * Async task to load AI component
 */
public class AILoaderTask extends AsyncTask<Void, Integer, Boolean> {

    /**
     * Context of the UI thread
     */
    Context context;

    /**
     * View of the UI thread
     */
    View rootView;

    /**
     * GameLogicTask constructor
     *
     * @param   context         Context of the UI thread
     * @param   rootView        View of the UI thread
     */
    public AILoaderTask(Context context, View rootView){
        this.context = context;
        this.rootView = rootView;
    }

    /**
     * Loads and sets the Neural Network of PlayerLogic
     *
     * @param   params      none
     * @return 	Boolean 	True when AI component loading is finished
     */
    @Override
    protected Boolean doInBackground(Void... params) {

        InputStream inStream = context.getResources().openRawResource(R.raw.player_neural_network);

        try {
            MultiLayerNetwork network = ModelSerializer.restoreMultiLayerNetwork(inStream);
            PlayerLogic.SetNeuralNetwork(network);
        } catch (IOException e) {
            e.printStackTrace();
        }

        return true;

    }

    /**
     * Accesses the UI thread - notifies the user that AI loading was successful
     *
     * @param   result      Result of doInBackground
     */
    @Override
    protected void onPostExecute(Boolean result) {
        Button btn = rootView.findViewById(R.id.buttonNewGame);
        btn.setText(R.string.button_new_game);
        btn.setEnabled(true);
        Snackbar.make(rootView, hu.bme.aut.android.chainreaction.R.string.ai_loaded, Snackbar.LENGTH_LONG).show();
    }

    @Override
    protected void onPreExecute() {
            Button btn = rootView.findViewById(R.id.buttonNewGame);
            btn.setText(R.string.button_loading);
            btn.setEnabled(false);
    }

    @Override
    protected void onProgressUpdate(Integer... values) {}

}

File: app/src/main/java/presenter/task/AILoaderTask.kt
package presenter.task

import android.support.design.widget.Snackbar
import android.view.View
import android.widget.Button
import model.ai.PlayerLogic
import android.content.Context
import android.os.AsyncTask
import hu.bme.aut.android.chainreaction.R
import org.deeplearning4j.util.ModelSerializer
import presenter.IMainView

import java.io.IOException

/**
 * Async task to load AI component
 */
class AILoaderTask
/**
 * GameLogicTask constructor
 *
 * @param   context         Context of the UI thread
 */
    (
    /**
     * The caller activity
     */
    val view: IMainView,
    /**
     * Context of the UI thread
     */
    val context: Context
) : AsyncTask<Void, Int, Boolean>() {

    /**
     * Loads and sets the Neural Network of PlayerLogic
     *
     * @param   params      none
     * @return  Boolean 	True when AI component loading is finished
     */
    override fun doInBackground(vararg params: Void): Boolean? {

        val inStream = context.resources.openRawResource(R.raw.player_neural_network)

        try {
            val network = ModelSerializer.restoreMultiLayerNetwork(inStream)
            PlayerLogic.SetNeuralNetwork(network)
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        }

        return true

    }

    /**
     * Accesses the UI thread - notifies the user about AI loading result
     *
     * @param   result      Result of doInBackground
     */
    override fun onPostExecute(result: Boolean) {
        view.loadedAI(result)
    }

    override fun onPreExecute() {
        view.notLoadedAI()
    }

    override fun onProgressUpdate(vararg values: Int?) {}

}
